cirq-google/cirq_google/ops/internal_gate_test.py

```diff
@@ -67,3 +71,74 @@ def test_internal_gate_with_hashable_args_is_hashable():
     )
     with pytest.raises(TypeError, match="unhashable"):
         _ = hash(unhashable)
+
+
+def test_internal_gate_with_custom_function_repr():
+    original_func = lambda x: x**2
+    x = np.linspace(-1, 1, 10)
+    y = original_func(x)
+    encoded_func = internal_gate.encode_function(x=x, y=y)
+
+    gate = internal_gate.InternalGate(
+        gate_name='GateWithFunction',
+        gate_module='test',
+        num_qubits=2,
+        custom_args={'func': encoded_func},
+    )
+
+    assert repr(gate) == (
+        "cirq_google.InternalGate(gate_name='GateWithFunction', "
+        f"gate_module='test', num_qubits=2, custom_args={str(gate.custom_args)})"
+    )
+
+    assert str(gate) == (f"test.GateWithFunction(func={encoded_func})")
+
+    with pytest.raises(ValueError):
+        _ = cirq.to_json(gate)
+
+
+def test_internal_gate_with_custom_function_round_trip():
+    original_func = lambda x: x**2
+    x = np.linspace(-1, 1, 10)
+    y = original_func(x)
+    encoded_func = internal_gate.encode_function(x=x, y=y)
+
+    gate = internal_gate.InternalGate(
+        gate_name='GateWithFunction',
+        gate_module='test',
+        num_qubits=2,
+        custom_args={'func': encoded_func},
+    )
+
+    msg = arg_func_langs.internal_gate_arg_to_proto(gate)
+
+    new_gate = arg_func_langs.internal_gate_from_proto(msg, arg_func_langs.MOST_PERMISSIVE_LANGUAGE)
+
+    func_proto = new_gate.custom_args['func'].function_interpolation_data
+
+    decoded_func = lambda x: np.interp(x, func_proto.independent_var, func_proto.dependent_var)
+
+    # Test original points evaluate to same value.
+    np.testing.assert_allclose(decoded_func(x), y)
+
+    # Test interpolation is close enough
+    eps = 1e-3
+    x_new = x[:-1] + eps
+    np.testing.assert_allclose(decoded_func(x_new), original_func(x_new), atol=2 * eps)
```

Comment:
This in effect tests if `np.interp` works, but that is not our scope.  Let us just test our code only -

```suggestion
    np.testing.assert_allclose(x, func_proto.independent_var)
    np.testing.assert_allclose(y, func_proto.dependent_var)
```